Mt7921e devel - MTK internal SQC fixes - v2

objelf opened the pull request:
Hi Lore, 

Please help review if there is any problem.

If the patch is fine to you, then I would add your Reviewed-by: tag and send them out to the upstream.

thanks,
Sean

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac.h:
```diff
@@ -134,8 +134,12 @@ mt76_connac_pm_unref(struct mt76_connac_pm *pm)
+	struct mt76_dev *dev = phy->dev;
 	bool ret;
 
+	if (mt76_is_mmio(dev) && dev->token_count)
```
@objelf I do not think we need the mt76_is_mmio() check is token_count is increased/decreased by mmio device only. I will remove the check merging the patch

objelf replied:
ok, thanks.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:
```diff
@@ -875,10 +875,13 @@ int mt76_connac_mcu_add_sta_cmd(struct mt76_phy *phy,
-		mt76_connac_mcu_sta_tlv(phy, skb, info->sta, info->vif,
-					info->rcpi);
+	if (info->sta) {
+		mt76_connac_mcu_sta_basic_tlv(skb, info->vif, info->sta,
```
@objelf mt76_connac_mcu_add_sta_cmd() is run by mt7663 driver with info->sta = NULL and info->enable = NULL in mt7615_remove_interface(). Can you please double check the mt76_connac_mcu_sta_basic_tlv() is not necessary in this case (mt7663 device)?

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:
```diff
@@ -1318,8 +1319,21 @@ int mt76_connac_mcu_uni_add_bss(struct mt76_phy *phy,
+				.len = cpu_to_le16(sizeof(struct bss_info_uni_bss_color)),
+				.enable = 0,
+				.bss_color = 0,
+			},
```
@objelf can you please confirm mt7663 just ignore this field?

objelf replied:
the UNI_BSS_INFO_BSS_COLOR only supports 80211ax chipset

objelf replied:
the UNI_BSS_INFO_BSS_COLOR only supports 80211ax chipset

LorenzoBianconi replied:
what I mean is, is bss_info_uni_bss_color element just skipped by 7663 fw or does it report an error? in the first case the patch is fine

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:
```diff
@@ -218,7 +218,7 @@ void mt76_connac_mcu_beacon_loss_iter(void *priv, u8 *mac,
 		return;
 
-	ieee80211_beacon_loss(vif);
+	ieee80211_connection_loss(vif);
```
@objelf is it true for mt7663 fw as well?

objelf replied:
The change should be dropped. let's focus on mt7921 first.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c:
```diff
@@ -13,6 +13,9 @@ int mt76_connac_pm_wake(struct mt76_phy *phy, struct mt76_connac_pm *pm)
+	if (test_bit(MT76_STATE_SUSPEND, &phy->state))
+		return 0;
+
 	cancel_delayed_work_sync(&pm->ps_work);
```
@objelf is it enough to just return after cancel_delayed_work_sync() if the device is not running (MT76_STATE_RUNNING not set)? Doing so we can drop the MT76_STATE_WOW patch, what do you think?

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/mac.c:
```diff
@@ -1301,24 +1302,30 @@ mt7921_mac_reset(struct mt7921_dev *dev)
+		clear_bit(MT76_STATE_PM, &dev->mphy.state);
+		dev->pm.stats.last_wake_event = jiffies;
+		dev->pm.stats.last_doze_event = jiffies;
+
```
@objelf why not moving dev->pm.stats.last_wake_event/dev->pm.stats.last_doze_event initialization and clear_bit() inside __mt7921_mcu_drv_pmctrl()?

objelf replied:
it seems doable. let's do it in the next version.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/mac.c:
```diff
@@ -1339,13 +1346,16 @@ void mt7921_mac_reset_work(struct work_struct *work)
 
-	queue_work(dev->mt76.wq, &dev->reset_work);
+	if (!test_bit(MT76_HW_RESET, &dev->mphy.state))
+		queue_work(dev->mt76.wq, &dev->reset_work);
```
@objelf do we really need this check? A similar order is already performed by queue_work() and mt76 is an ordered workqueue

LorenzoBianconi commented:
> Hi Lore,
> 
> Please help review if there is any problem.
> 
> If the patch is fine to you, then I would add your Reviewed-by: tag and send them out to the upstream.
> 
> thanks,
> Sean

Hi Sean, 

I merged most the patches and I added some comments in the other ones (most of them are trivial to fix)

Regards,
Lorenzo

objelf committed:
mt76: mt7921: add back connection monitor support

Hw beacon cmd to the mt7921 firmware doesn't only filter out the beacon,
but also performs its own connection monitoring, including periodic
keep-alives to the AP and probing the AP on beacon loss. Will indicate
the host with the event when the firmware detects the connection is lost.

Fixes: 1d8efc741df8 ("mt76: mt7921: introduce Runtime PM support")
Signed-off-by: Deren Wu <deren.wu@mediatek.com>
Signed-off-by: YN Chen <yn.chen@mediatek.com>
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

objelf commented:
close in order to send  the connection monitor patch
